returnn/torch/engine.py

```diff
@@ -259,6 +259,50 @@ def _load_model(self, epoch):
             print("Load model %s" % (filename,), file=log.v4)
             model_state = torch.load(filename)
             self._model.load_state_dict(model_state)
+        preload_from_files = self.config.typed_value("preload_from_files", {})
+        if preload_from_files:
+            # see `preload_from_files` in tf engine and `returnn.tf.network.CustomCheckpointLoader`
+            is_training = self.config.value("task", "train") == "train"
+            is_first_train_epoch = epoch == 1 and (
+                is_training or self.config.value("task", "train") == "initialize_model"
+            )
+            # We use the reversed sorted order here to achieve consistent behavior with the TF engine.
+            # There, the keys are used in sorted order but if a variable is loaded,
+            # it will not be considered anymore afterwards.
+            # So the first occurrence is used.
+            # Here, we overwrite variables even if they have been loaded before.
+            # In order to get consistent behavior, we use the reversed order.
+            for preload_key, opts in reversed(sorted(preload_from_files.items())):
+                assert isinstance(opts, dict) and "filename" in opts
+                if opts.get("init_for_train", False):
+                    if not is_first_train_epoch:
+                        continue
+                else:  # default: init for recog
+                    if is_training:
+                        continue
+                print(f"Pre-load weights for key '{preload_key}' from {opts['filename']}", file=log.v3)
+                preload_model_state = torch.load(opts["filename"])
+                if opts.get("checkpoint_key", None) is not None:
+                    preload_model_state = preload_model_state[opts["checkpoint_key"]]
+                if opts.get("prefix", ""):
```

Comment:
In the TF backend the `"prefix"` option gets passed to `load_if_prefix` of `CustomCheckpointLoader`, so the inverse option to `"ignore_prefix"`. What you implement here is something different, right? You just add the prefix to all variables. I don't think this is useful? Anyways should not have a different meaning than for the TF backend.

Reply:
I also had to look how it is done `CustomCheckpointLoader` because I always keep forgetting what `prefix` means. But as I understand the TF code, it expects that you have some params in the model named like `<prefix>paramN` or so, and then in the checkpoint file it expects it to be as `paramN` (without the prefix).

And if I understand the PT code correctly here, it should have the same effect, right?

Reply:
The model was usually prepared in a way that it has a subnetwork in `output/lm` or so for an external LM, or maybe it does model combination where there was a complete separate model, all with prefix `model2_`, and then you would load an existing checkpoint, but in the checkpoint it would be without this predix (`output/lm/` or `model2_`).

Reply:
Oof, I see. I would find it easier to understand if you get the list of model parameters first (`model.named_parameters()`?) and then select the parameters from `preload_model_state` that appear in the model parameters with the prefix.
And then also `ignore_params` and `ignore_params_prefixes` could be checked against the model parameters directly rather than against the altered checkpoint parameter names.

As it is now aren't valid parameters that are both in the checkpoint and the model, but just don't have the "prefix" printed as `unexpected_keys`? Even with the prefix which they never had.

Reply:
You mean you suggest to reimplement this, in a way you outlined?

Currently this code uses `load_state_dict`, but what you are suggesting cannot really use `load_state_dict` then, but have to replicate that logic. As I remember when looking at `load_state_dict`, it does a bit more than that, though. See code [here](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.load_state_dict) and then also `_load_from_state_dict`. It is also about persistent buffers, `set_extra_state`, pre-hooks, and maybe more.

Reply:
> it would be the same, or not?

Yes.
Except for the `unexpected_keys` message. Or do we expect that if "prefix" is used all parameters from the checkpoint will be present in the "prefix subnetwork" in all cases?

Reply:
Sorry, I didn't get the thing about `load_state_dict`. Of course the dict you pass to it would use the model parameter names, so with prefix. 🤔

Reply:
You're right about the unexpected keys. I added the prints because it seemed handy to me, but we can think about whether this should be modified. Maybe it could be combined for all contents of `preload_from_files` because if a key is missing in the first model but is loaded in the second, it could be misleading if the log shows it's missing in the first place. I often find the log output of the tf `preload_from_files` confusing, so not sure how much it can serve as a guidance.

Reply:
You mean a TODO comment that we want to improve the output?
Anyways, as I see now the behaviour, also in terms of logging, should be like the TF backend. So good to merge.

Reply:
No, I meant a comment describing what `prefix` actually does, and why we implement it this way, why we use `load_state_dict`.